### PR TITLE
[NeoML]TiedEmbedding access through composite layer

### PR DESCRIPTION
**Pull request by daniyalaliev**
TiedEmbedding now can access MultichannelLookupLayer in CompositeLayer
If lookUpLayer is inside of composite1/composite2 
then you can set SetEmbeddingsLayerName({ "composite1", "composite2 ", "lookUpLayer " })
Also can be used as usual (in case of lookup located not inside of composite)
SetEmbeddingsLayerName( "lookUpLayer " )

### CODE REVIEW

File: NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h [hunk 1]
```diff
@@ -33,8 +33,12 @@ class NEOML_API CTiedEmbeddingsLayer : public CBaseLayer {
 	
 	// Embeddings layer name from which we take the matrix.
 	// Only CMultichannelLookupLayer is supported.
-	const char* GetEmbeddingsLayerName() const { return embeddingsLayerName; }
-	void SetEmbeddingsLayerName( const char* name ) { embeddingsLayerName = name; }
+	// Storaged as path with size 1
+	const char* GetEmbeddingsLayerName() const { return embeddingPath.Last(); }
```
Comment (favorart) [inline NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h:39]: Fix comment
"// Embeddings layer name from which we take the matrix.
 // Only CMultichannelLookupLayer is supported.
 // Use this method if the lookupLayer is in the same level of the dnn (in the same composite layer)
 "

File: NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h [hunk 2]
```diff
@@ -33,8 +33,12 @@ class NEOML_API CTiedEmbeddingsLayer : public CBaseLayer {
...
+	void SetEmbeddingsLayerName(const char* name) { embeddingPath = { name }; }
+	// Full path to lookup layer if it is inside of composite
+	const CArray<CString>& GetEmbeddingsLayerPath() const { return embeddingPath; }
+	void SetEmbeddingsLayerPath(const CArray<CString>& path) { path.CopyTo(embeddingPath); }
```
Comment (favorart) [inline NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h:46]: Fix comment
"// Embeddings layer name from which we take the matrix.
 // Only CMultichannelLookupLayer is supported.
 // Use this method if the lookupLayer is in the nested level of the dnn (in some nested composite layer)
 "

File: NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp [hunk 3]
```diff
@@ -36,29 +36,35 @@ void CTiedEmbeddingsLayer::SetChannelIndex( int val )
...
 	CBaseLayer::Serialize( archive );
 
-	archive.Serialize( embeddingsLayerName );
+	if( version < 2001 ) {
+		if( archive.IsLoading() ) {
+			CString embeddingLayerName;
+			archive.Serialize( embeddingLayerName );
+			embeddingPath = { embeddingLayerName };
+		} else {
+			archive.Serialize( embeddingPath.Last() );
```
Comment (favorart) [inline NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp]: By storing serialize always the embeddingPath.
Check for a version only for the loading.

File: NeoML/src/Dnn/Dnn.cpp [hunk 4]
```diff
@@ -452,6 +452,27 @@ CPtr<const CBaseLayer> CDnn::GetLayer( const char* name ) const
 	return layerMap.Get( name );
 }
 
+CPtr<CBaseLayer> CDnn::GetLayer( const CArray<CString>& path)
+{
```
Comment (masepi) [inline NeoML/src/Dnn/Dnn.cpp:456]: AV in case of empty path

File: NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp [hunk 5]
```diff
@@ -16,8 +16,8 @@ limitations under the License.
 #include <common.h>
 #pragma hdrstop
 
-#include <NeoML/Dnn/Layers/TiedEmbeddingsLayer.h>
```
Comment (favorart) [inline NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp]: Update year of head comment in each file.

Revert this change of includes.

File: NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp [hunk 6]
```diff
@@ -152,9 +154,15 @@ const CDnnBlob* CTiedEmbeddingsLayer::getEmbeddingsTable() const
...
-	return embeddingsLayer->GetEmbeddings( channelIndex );
+	return getLookUpLayer()->GetEmbeddings( channelIndex );
+}
+
+CMultichannelLookupLayer* CTiedEmbeddingsLayer::getLookUpLayer() const
+{
+	CMultichannelLookupLayer* embeddingsLayer;
+	embeddingsLayer = CheckCast<CMultichannelLookupLayer>(
+		const_cast<CDnn*>(GetDnn())->GetLayer(embeddingPath).Ptr());
+	return embeddingsLayer;
```
Comment (favorart) [inline NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp:165]: I think, it can be avoid using the `const_cast<CDnn*>`, If you return `const CMultichannelLookupLayer*`

File: NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h [hunk 7]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 
 #include <NeoML/NeoMLDefs.h>
 #include <NeoML/Dnn/Dnn.h>
+#include <NeoML/Dnn/Layers/MultichannelLookupLayer.h>
```
Comment (favorart) [inline NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h]: Use forward declaration `class CMultichannelLookupLayer;` instead of include

File: NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp (hunk 6 above)
Comment (daniyalaliev) [reply NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp:165]: It shoudnt be const because it must change when called from LearnOnce

File: NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h [hunk 8]
```diff
@@ -20,6 +20,8 @@ limitations under the License.
```
Comment (favorart) [inline NeoML/include/NeoML/Dnn/Layers/TiedEmbeddingsLayer.h:20]: Fix for all changed files
from
/* Copyright © 2017-2020 ABBYY Production LLC
to
/* Copyright © 2017-2024 ABBYY

**Review by favorart — COMMENTED** [review]
ok

**Review by favorart — APPROVED** [review]
